src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs

```diff
@@ -0,0 +1,211 @@
+                Completing = 4,
+                Done = -2
+            }
+
+            internal AsyncEnumerable(ChannelReader<T> reader)
+            {
+                Debug.Assert(reader != null);
+
+                _reader = reader;
+                _state = (int)State.NotEnumerated;
+                _builder = AsyncIteratorMethodBuilder.Create();
+            }
+
+            public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken)
+            {
+                // Use this instance if it's never been enumerated before; otherwise, since it might still be in use,
+                // create a new instance.
+                AsyncEnumerable inst = Interlocked.CompareExchange(ref _state, (int)State.OuterLoop, (int)State.NotEnumerated) == (int)State.NotEnumerated ?
+                    this :
+                    new AsyncEnumerable(_reader) { _state = (int)State.OuterLoop };
+                inst._cancellationToken = cancellationToken;
+                return inst;
+            }
+
+            public ValueTask<bool> MoveNextAsync()
+            {
+                // Reset state for the next call.
+                _promise.Reset();
+                _usePromiseForResult = false;
+                _itemAvailable = false;
+
+                // Push the state machine forward.
+                MoveNext();
+
+                // If the operation completed asynchronously, return a ValueTask wrapping our promise-based implementation.
+                // If the operation completed synchronously, return a bool-backed ValueTask.
+                return _usePromiseForResult ?
+                    new ValueTask<bool>(this, _promise.Version) :
+                    new ValueTask<bool>(_itemAvailable);
+            }
+
+            public void MoveNext()
+            {
+                try
+                {
+                    bool waitToReadResult = false;
+                    switch ((State)_state)
+                    {
+                        case State.OuterLoop:
+                            ValueTask<bool> waitToReadTask = _reader.WaitToReadAsync(_cancellationToken);
+                            if (waitToReadTask.IsCompleted)
+                            {
+                                // WaitToReadAsync completed.  Get the result and jump to process it.
+                                waitToReadResult = waitToReadTask.GetAwaiter().GetResult();
+                                _state = (int)State.HaveWaitToReadResult;
+                                goto case State.HaveWaitToReadResult;
+                            }
+
+                            // WaitToReadAsync wasn't yet complete.  Mark that the promise
+                            // is being used for the result, set where we should return to when it completes,
+                            // store the awaiter, and hook up the continuation.
+                            _usePromiseForResult = true;
+                            _state = 1;
```

Comment:
> 1; [](start = 36, length = 3)

would it be better to use State.FinishingAwait?

Reply:
Oops, yes.